src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs

```diff
@@ -1310,53 +1312,237 @@ public static unsafe int LastIndexOfAny(ref byte searchSpace, byte value0, byte
         // Optimized byte-based SequenceEquals. The "length" parameter for this one is declared a nuint rather than int as we also use it for types other than byte
         // where the length can exceed 2Gb once scaled by sizeof(T).
         [MethodImpl(MethodImplOptions.AggressiveOptimization)]
-        public static unsafe bool SequenceEqual(ref byte first, ref byte second, nuint length)
+        public static bool SequenceEqual(ref byte first, ref byte second, nuint length)
         {
-            if (Unsafe.AreSame(ref first, ref second))
-                goto Equal;
+            nint offset = 0;
+            nint lengthToExamine = (nint)length;
 
-            IntPtr offset = (IntPtr)0; // Use IntPtr for arithmetic to avoid unnecessary 64->32->64 truncations
-            IntPtr lengthToExamine = (IntPtr)(void*)length;
+            // We do the small lengths early as the fixed costs are more significant for them.
+            if (length < sizeof(uint))
+            {
+                while (offset < lengthToExamine)
+                {
+                    if (Unsafe.Add(ref first, (IntPtr)offset) != (Unsafe.Add(ref second, (IntPtr)offset)))
```

Comment:
`nint` -> `IntPtr` casts are a performance trap. I believe that it will go to 64-bit long first on 32-platforms and then down-casted using checked cast to 32-bit again.

Reply:
Uses the `explicit operator IntPtr(int value)` -> `IntPtr(int value)` so should be ok? (rather than `nuint` which would go via `long`)

Reply:
FYI what Jan said is the reason the UTF-8 transcoding logic uses `void*` as an intermediary when converting between _IntPtr_ and _(whatever integral type)_.

https://github.com/dotnet/runtime/blob/a42c5839eed41e1a274eb90a84450e213cb65617/src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8Utility.Transcoding.cs#L114

Reply:
You are right. This should be fine. I thought there is unsigned/signed conversion too.

Reply:
But as @GrabYourPitchforks noted it is very easy to miss the cases where it is not fine. We had number of 32-bit specific perf bugs because of that.

Reply:
Yeah, I previously had an issue because pointers are unsigned so my less than zero tests always went the wrong way, which I hadn't expected :(